runner/scheduler.py

```diff
@@ -241,48 +245,72 @@ def pick_next_test_id(self, tests: dict[str, TestExample]) -> str:
                 msg = filler_no_response_tokens_trivia(remaining_tokens, self.agent.max_message_size)
                 tokens_spent = self.send_message("", SendMessageAction(msg, is_filling=True))
                 remaining_tokens -= tokens_spent
-            if not self.is_waiting(token_waiting_id, remove=True)[0]:
+            if not self.is_waiting(token_waiting_id, remove=True):
                 return token_waiting_id
 
         # As a last resort, if all tests are waiting for the percentage, get the closest one and force resume it.
         while True:
-            percentage_waiting_id = self.get_waiting_test("percentage_finished")
+            percentage_waiting_id = self.get_blocked_test("percentage_finished")
             if percentage_waiting_id is None:
                 break
             del self.wait_list[percentage_waiting_id]
             return percentage_waiting_id
 
         assert False, f"Couldn't find a test to run. Wait list: {self.wait_list}"
 
+    def fast_forward_tests(self, tests: dict[str, TestExample]):
+        # Go event by event in the master log and sync up with trace
+        finished_tests = {evt.test_id for evt in self.master_log.log if evt.type == EventType.END}
+        action = None
+        for evt in self.master_log.log:
+            if evt.type not in {EventType.BEGIN, EventType.SEND_MESSAGE, EventType.RESPONSE_MESSAGE, EventType.WAIT}:
+                continue
+            if evt.test_id in finished_tests:
+                continue
+            test = tests[evt.test_id]
+            self.wait_list.pop(evt.test_id, None)  # Since the test is performing an action, it can't be waiting.
+            match evt.type:
+                case EventType.BEGIN:
+                    result, skip = self.initialise_result(test)
+                    assert not skip
+                    self.in_progress_results[evt.test_id] = result
+                case EventType.SEND_MESSAGE:
+                    action = test.step()
+                    assert isinstance(action, SendMessageAction)
+                    assert action.message == evt.data["message"]
```

Comment:
This is exactly the problem that the Restaurant task had when resuming. This assert will likely fail on loading

Reply:
Good catch. In this case, responses are taken from the log, so the generator function has fewer chances to drift away from what was logged, but it could happen... Funny, it did not happen once in my tests, so I even forgot about the possibility. I'll go through it again and see what I can do. I remember thinking about implementing a sort of "forced mode" in the dynamic tests.

Reply:
Ideally, GPT would be deterministic and I wouldn't have to worry about these things, because any randomness in the dataset is also seeded 😅

Reply:
Actually, I think I answered myself. I will see if I can log the calls to `ask` from dynamic tests, so it can be loaded too.